pkg/instrumentation/flags.go

```diff
@@ -16,140 +16,218 @@ package instrumentation
+	case "disabled":
+		*s = Disabled
+	case "testing":
+		*s = Testing
+	case "production":
+		*s = Production
 	default:
-		return instrumentationError("invalid TraceConfig value of type %T: %v", obj, obj)
+		f, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return instrumentationError("invalid Sampling value '%s': %v", value, err)
+		}
+		*s = Sampling(f)
 	}
 	return nil
 }
 
-// Sampler returns a trace.Sampler corresponding to the TraceConfig value.
-func (tc TraceConfig) String() string {
+// String returns the Sampling value as a string.
+func (s Sampling) String() string {
 	switch {
-	case tc <= 0.005:
+	case s <= Disabled:
 		return "disabled"
-	case tc <= 0.1:
+	case s == Production:
 		return "production"
-	case tc >= 0.95:
+	case s >= Testing:
 		return "testing"
 	}
-	return strconv.FormatFloat(float64(tc), 'f', -1, 64)
+	return strconv.FormatFloat(float64(s), 'f', -1, 64)
 }
 
-// Sampler returns a trace.Sampler corresponding to the TraceConfig value.
-func (tc TraceConfig) Sampler() trace.Sampler {
+// Sampler returns a trace.Sampler corresponding to the Sampling value.
+func (s Sampling) Sampler() trace.Sampler {
 	switch {
-	case tc >= 0.95:
+	case s >= Testing:
 		return trace.AlwaysSample()
-	case tc <= 0.005:
+	case s <= Disabled:
 		return trace.NeverSample()
 	default:
-		return trace.ProbabilitySampler(float64(tc))
+		return trace.ProbabilitySampler(float64(s))
+	}
+}
+
+// parseEnv parses the environment for default values.
+func parseEnv(name, defval string, parsefn func(string) error) {
+	if envval := os.Getenv(name); envval != "" {
+		err := parsefn(envval)
+		if err == nil {
+			return
+		}
+		log.Error("invalid environment %s=%q: %v, using default %q", name, envval, err, defval)
+	}
+	if err := parsefn(defval); err != nil {
+		log.Error("invalid default %s=%q: %v", name, defval, err)
 	}
 }
 
 // defaultOptions returns a new options instance, all initialized to defaults.
 func defaultOptions() interface{} {
-	collector := os.Getenv("JAEGER_COLLECTOR")
-	agent := os.Getenv("JAEGER_AGENT")
-	metrics := os.Getenv("PROMETHEUS_ENDPOINT")
+	o := &options{}
 
-	if collector == "" {
-		collector = defaultCollector
+	type param struct {
+		defval  string
+		parsefn func(string) error
 	}
-	if agent == "" {
-		agent = defaultAgent
-	}
-	if metrics == "" {
-		metrics = defaultMetrics
+
+	params := map[string]param{
+		"JAEGER_COLLECTOR": {
+			defaultJaegerCollector,
+			func(v string) error { o.JaegerCollector = v; return nil },
+		},
+		"JAEGER_AGENT": {
+			defaultJaegerAgent,
+			func(v string) error { o.JaegerAgent = v; return nil },
+		},
+		"HTTP_ENDPOINT": {
+			defaultHTTPEndpoint,
+			func(v string) error { o.HTTPEndpoint = v; return nil },
+		},
+		"PROMETHEUS_EXPORT": {
+			defaultPrometheusExport,
+			func(v string) error {
+				enabled, err := utils.ParseEnabled(v)
+				if err != nil {
+					return err
+				}
+				o.PrometheusExport = enabled
+				return nil
+			},
+		},
+		"SAMPLING_FREQUENCY": {
```

Comment:
Why these are environment variables (as opposed to regular configuration values)?

Reply:
They are also configuration variables. IOW, you can set them in the configuration as well.
It seems to be a common pattern with Jaeger/Prometheus/OpenCensus that one can set parameters in the environment as well as on the command line. I'm not sure of the exact reason, but I guess the idea is to allow one to easily enable data collection for every single component on a node that supports it, without having to modify every single systemd service, etc/sysconfig, or whatever else file your distro would otherwise need to accomplish such a feat. That's also the reason these variables are not 'namespaced'. I tried to use the same names I saw in examples.